src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+        /// to not having enough permissions to query for the data of that specific process
+        /// </returns>
+        [DllImport(Interop.Libraries.libproc, SetLastError = true)]
+        private static unsafe extern int proc_pidinfo(
+            int pid,
+            int flavor,
+            ulong arg,
+            ulong* buffer,
+            int bufferSize);
+
+        /// <summary>
+        /// Gets the process information for a given process
+        /// </summary>
+        /// <param name="pid">The PID (process ID) of the process</param>
+        /// <returns>
+        /// Returns a valid proc_taskallinfo struct for valid processes that the caller 
+        /// has permission to access; otherwise, returns null
+        /// </returns>
+        internal static unsafe proc_taskallinfo? GetProcessInfoById(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Get the process information for the specified pid
+            int size = Marshal.SizeOf<proc_taskallinfo>();
+            proc_taskallinfo info = default(proc_taskallinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTASKALLINFO, 0, &info, size);
+            return (result == size ? new proc_taskallinfo?(info) : null);
+        }
+
+        /// <summary>
+        /// Gets the thread information for the given thread
+        /// </summary>
+        /// <param name="thread">The ID of the thread to query for information</param>
+        /// <returns>
+        /// Returns a valid proc_threadinfo struct for valid threads that the caller
+        /// has permissions to access; otherwise, returns null
+        /// </returns>
+        internal static unsafe proc_threadinfo? GetThreadInfoById(int pid, ulong thread)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Negative TIDs are invalid
+            if (thread < 0)
+            {
+                throw new ArgumentOutOfRangeException("thread");
+            }
+
+            // Get the thread information for the specified thread in the specified process
+            int size = Marshal.SizeOf<proc_threadinfo>();
+            proc_threadinfo info = default(proc_threadinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTHREADINFO, (ulong)thread, &info, size);
+            return (result == size ? new proc_threadinfo?(info) : null);
+        }
+
+        internal static unsafe List<KeyValuePair<ulong, proc_threadinfo?>> GetAllThreadsInProcess(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            int result = 0;
+            int size = 20; // start assuming 20 threads is enough
```

Comment:
The max number of threads per machine is set to 5x the max number of processes (based on an, admittedly outdated, doc I found on Apple's website https://support.apple.com/en-us/HT3854) and the per-process limit seems to be 2560 (per this thread on StackOverflow: http://stackoverflow.com/questions/2860889/cant-get-past-2542-threads-in-java-on-4gb-imac-osx-10-6-3-snow-leopard-32bit). Based on these limits and that any .NET, Java, or even Swift apps will have GC threads (among actual process threads), 20 seems like a safe number to make sure we get all threads for trivial apps. For bigger apps we'll need to increase that, but since we start with a reasonably high number, it should limit the number of times we need to double our buffer, giving a perf improvement for those apps. 

Thoughts?